arxiv/arxiv.py

```diff
@@ -233,12 +233,22 @@ def slugify(obj):
     return filename
 
 
-def download(obj, dirpath='./', slugify=slugify):
+def download(obj, dirpath='./', slugify=slugify, source=False):
+    """
+    Download the .pdf ob object. If source==True download the
+    source tar.gz instead
+    """
     if not obj.get('pdf_url', ''):
         print("Object has no PDF URL.")
         return
     if dirpath[-1] != '/':
         dirpath += '/'
-    path = dirpath + slugify(obj) + '.pdf'
-    urlretrieve(obj['pdf_url'], path)
+    if source:
+        url = re.sub(r'/pdf/', r'/e-print/',obj['pdf_url'])
+        path = dirpath + slugify(obj) + '.tar.gz'
```

Comment:
Major issue: this is not guaranteed to point to a tarfile; indeed, in many cases it will just be a PDF. From [arXiv's stock `/format` descriptions:](https://arxiv.org/format/2001.10010)

> **Source**
> Delivered as a gzipped tar (.tar.gz) file if there are multiple files, *otherwise as a PDF file*, or a gzipped TeX, DVI, PostScript or HTML (.gz, .dvi.gz, .ps.gz or .html.gz) file depending on submission format.

For example, `https://arxiv.org/e-print/2002.07816` is a tarfile; `https://arxiv.org/e-print/1707.08567v1` is a PDF.

The correct behavior here would be to select a filename based on the `Content-Type` of the e-print address. Two ways of doing this:

1. **Easy but slow:** first make a `HEAD` request to the e-print URL; if `application/x-eprint-tar`, proceed for a tarfile; else, proceed for a PDF.

2. **Fast, but with filesystem complexity:** `urlretrieve` the e-print URL without setting the destination path, which puts the file in a temporary directory. Look up the `Content-Type` header in the returned tuple, construct a content-type-dependent path, and move the file from its temporary location to that path.